### ISSUE TITLE
SchemaConfiguration file location and/or extension

### ISSUE THREAD

**Issue by wow-such-amazing** [apollographql/apollo-ios#3053]
### Use case

In our project we wanted to hide generated swift files as we didn't want to see them in PR-s so we decided to put put those generated files into Derived Data folder. 

This has resulted into SchemaConfiguration files being removed each time we do a Clean Derived Data. Which makes sense 😄 So our solution was to put SchemaConfiguration files inside the main project and adjust our code generation script to remove SchemaConfiguration files from Derived Data after each generation.

### Describe the solution you'd like

It would be great to have those possibilities to mitigate such issues:
- Be able to specify location for SchemaConfiguration files
- Be able to use extensions on SchemaConfiguration types

Cheers!

**Comment by calvincestari** [comment]
Hi @wow-such-amazing - could you not use gitignore for the generated files folder with exceptions specifically for the schema configuration + custom scalars files? That way they're excluded from source control (won't appear in PRs) and you don't have to use different locations/extensions for them.

```
MyProject/GeneratedFiles/
!MyProject/GeneratedFiles/**/SchemaConfiguration.swift
```

We have no plans to support separate output locations for schema configuration files.

**Comment by calvincestari** [comment]
Yes I know. I'm saying you don't need to. Storing the files in Derived Data is creating this issue when there is another way to achieve what you want.

Is there a particular reason the files must be in Derived Data that gitignore won't solve for you?

**Comment by shahrukh-sc** [comment]
Hey! @calvincestari 

I am with @wow-such-amazing in the team, we are using **Pre-Build Plugin** which uses **ApolloCodegen** to generate the Swift files

## what we wanted
- to ignore generated files so that it will not appear on PRs

## why use Pre-Build Plugin
- it should work as our earlier setup with Run-Script
- we shouldn't manually add the generated Source files to the Target

## what's the problem
- **Pre-Build Plugin** can only generate files in **DerivedData**
- `SchemaConfiguration.swift` file is also generated in the same **Generated** folder in **DerivedData**
- But, we don't want it to be away from your eyes, we want to override & commit this file to our repo

## what's our temporary solution
- we manually remove `SchemaConfiguration.swift` file from the **Generated** folder in **DerivedData** every time **ApolloCodegen** is run
- we manually added `SchemaConfiguration.swift` to the Target & overridden & committed it

## what's our proposal/request
- since `SchemaConfiguration.swift` directly impacts how Apollo controls cache
  - is it possible to specify location for `SchemaConfiguration.swift`?
  - or override the implementation of `cacheKeyInfo(for:, object:)` in an extension or something
  - or not generate this file at all by providing some argument to **ApolloCodegen**?

I hope it gives a clear idea what we are trying to do, please suggest if you have a better way to achieve the same 🙏

**Comment by calvincestari** [comment]
Hi @shahrukh-sc, thanks for the extra detail.

> it should work as our earlier setup with Run Phase Script

You can still use 1.0 code generation in a Run Phase script but you will probably have to use the CLI tool, instead of calling code generation through code. We removed the recommendation of using a Run Phase script when we released 1.0 because code generation typically does not need to happen on every build.

> we shouldn't manually add the generated Source files to the Target

This is a result of the code generation configuration and the `output` configuration parameter. It has support for a wide range of module types, some automated and some manual.

> **Pre-Build Plugin** can only generate files in **DerivedData**

Honestly, I think this is the limitation in this setup. If the plugin is only doing codegen then you might be able to drop it and go back to using a Run Phase script that uses the CLI directly instead. The legacy help documentation has [example scripts](https://www.apollographql.com/docs/ios/v0-legacy/installation#5-add-a-code-generation-build-step) that generate code to the project root instead of derived data. If the plugin is doing more than just codegen then you may not be able to, it's difficult to say without knowing more about the plugin.

> since SchemaConfiguration.swift directly impacts how Apollo controls cache
> * is it possible to specify location for SchemaConfiguration.swift?
> * or override the implementation of cacheKeyInfo(for:, object:) in an extension or something
> * or not generate this file at all by providing some argument to ApolloCodegen?

No, the schema configuration file is part of the schema module and generated into the same location as the other schema module files. The schema configuration file is the override of the `cacheKeyInfo(for:, object:)` function. There is no other way to do that.

> I hope it gives a clear idea what we are trying to do, please suggest if you have a better way to achieve the same

I still believe .gitignore is the better way to prevent files from appearing in SCM and subsequent PRs rather than generating files to an untracked folder outside of the repo.

Also note that you are only having this issue with the schema configuration file at the moment but as soon as you start using custom scalars you'll need to manually manage those files too since they're also not overwritten and can contain your own custom code for scalar transformations.

We may introduce new files or change others to follow the same pattern so you should know that it's a fragile workflow you have here. I'm confident that the current configuration options can suit a wide range of project configurations and workflows but yours is unfortunately outside of those cases. We have no plans to split location of the schema configuration files. You will need to fork the apollo-ios repo and change the file generation logic to have it work in the way you need for your project.

**Comment by shahrukh-sc** [comment]
Yeah, it looks like we have to change our approach to be future proof. Thanks much @calvincestari for the valuable inputs 💚
